File: esp_board_manager/devices/dev_knob/dev_knob.c
/*
 * SPDX-FileCopyrightText: 2026 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include <string.h>
#include "esp_board_device.h"
#include "esp_board_entry.h"
#include "esp_board_manager_defs.h"
#include "esp_err.h"
#include "esp_log.h"
#include "dev_knob.h"

static const char *TAG = "DEV_KNOB";

int dev_knob_init(void *cfg, int cfg_size, void **device_handle)
{
    if (cfg == NULL || device_handle == NULL ||
        cfg_size != sizeof(dev_knob_config_t)) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    const dev_knob_config_t *config = (const dev_knob_config_t *)cfg;
    if (config->sub_type == NULL) {
        ESP_LOGE(TAG, "Invalid knob sub_type");
        return -1;
    }

    const esp_board_entry_desc_t *entry_desc = esp_board_entry_find_subtype_desc(
        ESP_BOARD_DEVICE_TYPE_KNOB, config->sub_type);
    if (entry_desc == NULL || entry_desc->init_func == NULL) {
        ESP_LOGE(TAG, "Unsupported knob sub_type: %s", config->sub_type);
        return -1;
    }

    return entry_desc->init_func(cfg, cfg_size, device_handle);
}

int dev_knob_deinit(void *device_handle)
{
    if (device_handle == NULL) {
        ESP_LOGE(TAG, "Invalid device handle");
        return -1;
    }

    dev_knob_config_t *config = NULL;
    esp_err_t ret =
        esp_board_device_get_config_by_handle(device_handle, (void **)&config);
    if (ret != ESP_OK || config == NULL || config->sub_type == NULL) {
        ESP_LOGE(TAG, "Failed to get knob configuration: %s", esp_err_to_name(ret));
        return -1;
    }

    const esp_board_entry_desc_t *entry_desc = esp_board_entry_find_subtype_desc(
        ESP_BOARD_DEVICE_TYPE_KNOB, config->sub_type);
    if (entry_desc == NULL || entry_desc->deinit_func == NULL) {
        ESP_LOGE(TAG, "Unsupported knob sub_type: %s", config->sub_type);
        return -1;
    }

    ret = entry_desc->deinit_func(device_handle);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to deinitialize knob sub_type %s", config->sub_type);
    }
    return ret;
}

File: esp_board_manager/devices/dev_knob/dev_knob.py
# SPDX-FileCopyrightText: 2026 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

"""Knob device configuration parser."""

VERSION = 'v1.0.0'

_GPIO_SUB_TYPE = 'gpio'

DEV_KNOB_IO_LIST = {
    _GPIO_SUB_TYPE: ['gpio_encoder_a', 'gpio_encoder_b'],
}


def get_includes() -> list:
    """Return required include headers for the knob device."""
    return ['dev_knob.h']


def _get_gpio_number(config: dict, key: str) -> int:
    """Read and validate one encoder GPIO number."""
    if key not in config:
        raise ValueError(f'Knob device requires {key}')
    value = config[key]
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f'Knob device {key} must be an integer')
    gpio_num = value
    if gpio_num < 0:
        raise ValueError(f'Knob device {key} must be non-negative')
    if gpio_num > 255:
        raise ValueError(f'Knob device {key} must be between 0 and 255')
    return gpio_num


def parse(name: str, full_config: dict, peripherals_dict=None) -> dict:
    """Parse a GPIO quadrature knob configuration into a C initializer."""
    del peripherals_dict

    peripherals = full_config.get('peripherals', [])
    if peripherals is None:
        peripherals = []
    if not isinstance(peripherals, list):
        raise ValueError('Knob device peripherals must be a list')
    if peripherals:
        raise ValueError('Knob device does not support peripheral bindings')

    sub_type = full_config.get('sub_type')
    if sub_type != _GPIO_SUB_TYPE:
        raise ValueError("Knob device must use sub_type 'gpio'")

    config = full_config.get('config', {})
    if not isinstance(config, dict):
        raise ValueError('Knob device config must be a mapping')

    gpio_encoder_a = _get_gpio_number(config, 'gpio_encoder_a')
    gpio_encoder_b = _get_gpio_number(config, 'gpio_encoder_b')
    if gpio_encoder_a == gpio_encoder_b:
        raise ValueError('Knob device gpio_encoder_a and gpio_encoder_b must be different')

    default_direction = config.get('default_direction', 0)
    if isinstance(default_direction, bool) or not isinstance(default_direction, int):
        raise ValueError('Knob device default_direction must be an integer')
    if default_direction not in (0, 1):
        raise ValueError('Knob device default_direction must be 0 or 1')

    use_rtc = config.get('use_rtc', False)
    if not isinstance(use_rtc, bool):
        raise ValueError('Knob device use_rtc must be a boolean')

    enable_power_save = config.get('enable_power_save', False)
    if not isinstance(enable_power_save, bool):
        raise ValueError('Knob device enable_power_save must be a boolean')

    return {
        'struct_type': 'dev_knob_config_t',
        'struct_var': f"{name.replace('-', '_')}_cfg",
        'struct_init': {
            'name': name,
            'sub_type': sub_type,
            'use_rtc': use_rtc,
            'knob_config': {
                'default_direction': default_direction,
                'gpio_encoder_a': gpio_encoder_a,
                'gpio_encoder_b': gpio_encoder_b,
                'enable_power_save': enable_power_save,
            },
        },
    }
